contrib/cf-remote/cf_remote/packages.py

```diff
@@ -0,0 +1,188 @@
+from cf_remote.web import get_json
+from cf_remote.utils import pretty, is_in_past, canonify, file_extension, file_name
+import cf_remote.log as log
+import re
+
+
+def get_releases():
+    data = get_json("https://cfengine.com/release-data/enterprise/releases.json")
+    return data["releases"]
+
+
+def get_latest_releases():
+    releases = get_releases()
+    for release in releases:
+        if "status" in release and release["status"] == "unsupported":
+            continue
+        if "latest_on_branch" not in release or not release["latest_on_branch"]:
+            continue
+        yield release
+
+
+def get_latest_stable():
+    releases = get_releases()
+    for release in releases:
+        if "latest_stable" in release and release["latest_stable"]:
+            return release
+    return None
+
+
+def get_latest_stable_data():
+    return get_release_data(get_latest_stable())
+
+
+def get_release_data(release):
+    assert "URL" in release
+    return get_json(release["URL"])
+
+
+def get_default_version(edition=False):
+    numeric = get_latest_stable()["version"]
+    if edition:
+        return "CFEngine Enterprise " + numeric
+    return numeric
+
+
+def get_default_version_data():
+    return get_latest_stable_data()
+
+
+class Artifact:
+    def __init__(self, data, filename=None):
+        if filename and not data:
+            data = {}
+            data["URL"] = "./" + filename
+            data["Title"] = ""
+            data["Arch"] = None
+            if "sparc" in filename.lower():
+                data["Arch"] = "spar"
+        self.data = data
+        self.url = data["URL"]
+        self.title = data["Title"]
+        self.arch = canonify(data["Arch"])
+
+        self.file_name = file_name(self.url)
+        self.file_extension = file_extension(self.file_name)
+
+        self.tags = ["any"]
+
+        self.add_tag(self.arch)
+        self.add_tag(self.file_extension)
+
+        look_for_tags = [
+            "Windows", "CentOS", "Red Hat", "Debian", "Ubuntu", "SLES", "Solaris", "AIX", "HPUX",
+            "HP-UX", "nova", "community", "masterfiles", "OSX", "OS X", "OS-X"
+        ]
+        for tag in look_for_tags:
+            tag = tag.lower()
+            filename = self.file_name
+            if tag in filename:
+                self.add_tag(tag)
+        if "hp-ux" in self.tags:
+            self.add_tag("hpux")
+            self.tags.remove("hp-ux")
+        if "os-x" in self.tags:
+            self.add_tag("osx")
+            self.tags.remove("os-x")
+        if "os_x" in self.tags:
+            self.add_tag("osx")
+            self.tags.remove("os_x")
+
+        parts = re.split(r"[-_.]", self.file_name)
+        #parts = self.file_name.split("-_.")
+        for part in parts:
+            part = part.strip()
+            if part == "x86":
+                continue
+            try:
+                _ = int(part)
+            except ValueError:
+                self.add_tag(part)
```

Comment:
I think the *tag magic* above deserves a separate method.

Reply:
Added.